matching/README.md

```diff
@@ -0,0 +1,991 @@
+Maranget's paper that rows at the top of the matrix apply first if they can,
+and lower rows only apply if the upper rows cannot. Instead, any row in the
+topmost group can apply if it matches, and rows in lower groups only apply if
+no row in an upper group applies. This allows us to improve pattern matching
+efficiency since K does not, for the most part, case what order rules are tried
+in.
+
+Matrices are first simplified to remove or-patterns by replacing each row
+containing an or-pattern with a set of rows corresponding to all possible
+permutations of the disjuncts of all or-patterns in the pattern.
+
+### Decision Trees
+
+A decision tree is a tree whose root represents the input of the matching
+process and whose leaves identify which rewrite rule, if any, applies to the
+term being matched. It has the following BNF-style syntax:
+
+```
+syntax DT ::= "Failure"
+            | Success(Int, [(Occurrence, Sort)])
+            | Switch(Occurrence, Sort, [(Constructor, DT)], DT?)
+            | CheckNull(Occurrence, Sort, DT, DT)
+            | Function(Occurrence, Function, [(Occurrence, Sort)], Sort, DT)
+            | Pattern(Occurrence, Sort, Pattern, DT)
+            | IterHasNext(Occurrence, Sort, DT, DT)
+```
+
+### Heuristics
+
+The only heuristics regularly used by the LLVM backend are `q`, `b`, `a`, and
+the `L` pseudo-heuristic. We will discuss in a later section how these
+heuristics are adapted to the new features of the algorithm.
+
+## Compilation
+
+Here we describe the compilation algorithm, the updated specialization and
+default decompositions, and the adapted heuristics.
+
+### Initial matrix
+
+Starting from a set of rewrite rules of size n and their priorities, where each
+rewrite rule has a left-hand side and a right-hand side of sort s, and
+optionally a boolean condition Cond, we compute the initial matrix as follows:
+
+P starts with one column and n rows. P_i1 = LHS[i]. o_1 = []. s_1 = s.
+C_i = (a_i = i, S_i = empty, Cond_i = Cond[i], R_i = []).
+
+The rows of P are sorted in increasing priority order, where each set of rows
+with the same priority forms a group.
+
+### Compile function
+
+`Compile` is a function that takes a Matrix and returns a Decision Tree.
+
+Described below is pseudocode for the function:
+
+```
+def compile(Matrix m): DT
+  if m has no rows:
+    return Failure
+  else:
+    case bestRow(m) of
+      None -> decomposeBestCol(m)
+      BestRow(i) -> getLeaf(i, compile(m with row i removed))
+
+def bestRow(Matrix m): Int
+  if m has no columns:
+    return the first row of m
+  if there is a row in the topmost group of m where all columns are wildcard patterns:
+    return that row
+  else:
+    return None
+
+def getLeaf(Row i, DT dt): DT
+  val newVars = empty
+  for j = 1 to m:
+    for each variable X in P_ij:
+      newVars[X] = o_j
+  val allVars = newVars U S_i
+  val vars = sort the values of allVars by their keys alphabetically
+  val atomicLeaf = Success(a_i, vars)
```

Comment:
How are the fringe entries `o_j` initialised? On line 137 you set up the first entry `o_1 = []`, but don't mention further cases. What makes `o_1` different such that it's initialised explicitly and each further `o_i` is implicitly pulled in from the term being matched?

Here, if `P_i = [x(A, B) y(C) D]` then we'll end up with:
```
newVars[A] = o_1
newVars[B] = o_1
newVars[C] = o_2
newVars[D] = o_3
```
I don't follow why we'd want `A` and `B` to map to the same occurrence. Wouldn't we actually want to know that `A` maps to the occurrence `o_1 . 1` (i.e. the first child of `o_1`):
```
newVars[A] = o_1 . 1
newVars[B] = o_1 . 2
newVars[C] = o_2 . 1
newVars[D] = o_3
```

Reply:
I added the following text to the section `Matrices` in the preliminaries and I hope it makes the notation clearer:

```The fringe consists of m occurrences and m sorts, one for each
column. We call these occurrences for convenience o_1, ..., o_m and the sorts
s_1, ..., s_m. What actual occurrence corresponds to o_1 depends on what
occurrence is in the first column of the matrix and will change as the matrix
transforms```